src/controllers/hid/legacyhidcontrollermapping.h

```diff
@@ -13,8 +12,11 @@ class LegacyHidControllerMapping : public LegacyControllerMapping {
     ~LegacyHidControllerMapping() override {
     }
 
+    LegacyControllerMapping* clone() const override {
```

Comment:
Consistently use `LegacyControllerMappingPointer` everywhere, not plain pointers. Your code already has memory leaks.

Change the typedef from QSharedPointer to std::shared_pointer and then you could use `dynamic_pointer_cast` instead of `dynamic_cast`.

http://www.cplusplus.com/reference/memory/dynamic_pointer_cast/

Alternatively use `qSharedPointerDynamicCast` but I would prefer the standard C++ route instead of dealing with custom Qt re-inventions if not strictly necessary.

Reply:
I'll get rid of the QSharedPointer and use std. Might make the diff quite big though.

Reply:
If the diff is too large then keep using QSharedPointer. Doesn't really matter.

Reply:
Come to think of it, std::unique_ptr might be a better choice here.

Reply:
Ugh, there's no equivalent of `dynamic_pointer_cast` for unique_ptr...

Reply:
Well, there's this abomination:
```
    m_pMapping = std::unique_ptr<LegacyMidiControllerMapping>(
        dynamic_cast<LegacyMidiControllerMapping*>(pMapping.release()));
```

Reply:
Did think about this, too. But then discarded the thought to not confuse you ;) One step at a time. It doesn't really matter, though std::unique_ptr ensures mappings need to be cloned and can't be shared. Depends on the use case.

Reply:
The whole point of the cloning is to ensure it is not shared, so I think unique_ptr is a better solution.

Reply:
This is a potential memory leak. The dynamic_cast may nullify a valid pointer without getting hold of it again! Use get() immediately followed by release() if valid. Very ugly and error prone.

Reply:
Like this?

```
void HidController::setMapping(std::unique_ptr<LegacyControllerMapping> pMapping) {
    m_pMapping = std::unique_ptr<LegacyHidControllerMapping>(
        dynamic_cast<LegacyHidControllerMapping*>(pMapping.get()));
    if (m_pMapping) {
        pMapping.release();
    }
}
```

Reply:
Maybe add a comment: // Downcast and take ownership on success

It's a shame that you need to write so much code **and** a comment to get this right. C++ :weary:

Reply:
:see_no_evil:

Reply:
Actually, a shared pointer is needed. DlgPrefController, ControllerInputMappingTableModel, and ControllerOutputMappingTableModel all share a pointer in the main thread.

Reply:
```
void HidController::setMapping(std::unique_ptr<LegacyControllerMapping> pMapping) {
    m_pMapping = std::unique_ptr<LegacyHidControllerMapping>(
        dynamic_cast<LegacyHidControllerMapping*>(pMapping.get()));
    if (m_pMapping) {
        pMapping.release();
    }
}
```
This seems to double free or something? Forget it, I'll use shared pointers and DEBUG_ASSERT that use_count is 1. :see_no_evil:

Reply:
Exactly oncy. It is freed later by `m_pMapping` if the dynamic_cast succeeded or immediately when `pMapping` goes out of scope. release() doesn't free.